rir/src/compiler/analysis/visibility.cpp

```diff
@@ -30,9 +30,16 @@ AbstractResult VisibilityAnalysis::apply(LastVisibilityUpdate& vis,
 
         if (builtinUpdatesVisibility(builtinId)) {
             isVisibilityChanging();
+            break;
         }
-        break;
+    // fall through
     default:
+        // This instruction might change visibility, thus it's visibility effect
```

Comment:
It might be simpler to just add `Visible` and `Invisible` to `observable` and not have `last`, I don't see where `last` is used any different than `observable`. Then you would also put this outside of the default case.

Also line 11, `if (vis.last != i) {`, wouldn't that always be false, or when would we update analysis on the same instruction twice?

Reply:
thanks for the excellent suggestions